<a href="https://colab.research.google.com/github/daria-pavlova/python_for_ds_tasks/blob/main/hw_2.1_1_hypothesis_testing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

Імопортуємо метод для проведення ztest.

In [6]:
import numpy as np
from statsmodels.stats.weightstats import ztest

Задаємо нашу вибірку зі зростом студентів.

In [1]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166, 173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

Задаємо рівень значущості та нульову гіпотезу середнього зросту.

In [7]:
alpha = 0.05
null_height = 170

Розраховуємо середнє значення вибірки та середнє відхилення від середнього значення.

In [9]:
print('mean=%.2f stdv=%.2f' % (np.mean(heights), np.std(heights)))

mean=170.70 stdv=4.65


Тепер проводимо тест. У цій функції ми передали дані, у параметрі значення ми передали середнє значення в нульовій гіпотезі, в альтернативній гіпотезі ми перевіряємо, чи більше середнє значення.

In [10]:
ztest_Score, p_value = ztest(heights, value = null_height, alternative = 'larger')

In [11]:
ztest_Score, p_value

(0.8482235371587619, 0.19815674571091746)

Дивимося, яку з гіпотез задовільняє наше p value.

In [13]:
if (p_value < alpha):
  print("Reject H0. Mean height is less than 170.")
else:
  print("H0 can't be rejected. Mean height is greater than 170.")

H0 can't be rejected. Mean height is greater than 170


### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [15]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

Перевіряємо розміри вибірок, щоб впевнитися, що вони більше 30.

In [16]:
len(group_a_scores), len(group_b_scores)

(40, 40)

Рораховуємо середнє значення для кожної з вибірок.

In [17]:
np.mean(group_a_scores), np.mean(group_b_scores)

(77.08125, 74.73875000000001)

Виконуємо z-test.

In [19]:
ztest(group_b_scores, group_a_scores, value=0)

(-2.357354864751502, 0.01840564855252801)

Статистика між двома вибірках по z-критерію становить 2.3573, відповідно p-value становить 0.0184.
Оскільки p-value < 0.05, у нас достатньо доказів, щоб відкинути нульову теорію. Іншими словами оцінки учнів вибірок значно різняться між 2 вибірками.
Додатково ми можемо зробити висновок, що стандартна девіація групи А негативна - 2.3573, тобто спостерігається погіршення оцінок на основі нового курсу.

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

Обрізаємо оригінальний масив вибірки групи А і зберігаємо його. в нову вибірку для статистичного аналалізу t-test.

In [22]:
group_a_scores_only_five = group_a_scores[:5]
group_a_scores_only_five

[78.55, 72.25, 79.88, 75.0, 76.54]

Імпортуємо t-test.

In [28]:
from scipy.stats import ttest_ind

Переглянемо середні значення для двох вибірок.

In [30]:
np.mean(group_a_scores_only_five), np.mean(group_b_scores)

(76.444, 74.73875000000001)

Проведемо t-test, щоб визначити девіацію між вибірками.

In [34]:
t_stat, p_val = ttest_ind(group_b_scores, group_a_scores_only_five)

print(f'T-statistic is {t_stat}')
print(f'P-value is {p_val}')

T-statistic is -0.8167951030410814
P-value is 0.4185479477535733


Знову можемо побачити негативні результати вже значення t-statistic -0.8167, p-value is > 0.05, therefore we fail to reject H0. Which means that results between 2 data groups has no sufficient evidence to say that the difference between 2 groups is statisticly relevant.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [36]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [42]:
diff = np.array(after_training) - np.array(before_training)

In [43]:
z_stat, p_val = ztest(diff, value=0)

print(f'Z-statistic: {z_stat}')
print(f'p-value: {p_val}')

Z-statistic: 9.13890009623916
p-value: 6.309071983835101e-20


Результат свідчить про те, p-value > 0.05. Тобто ми можемо стверджувати, що різниця мід двома вибірками є статистино значущою.
Результати z-test позитивний, що свідчиить про те, що тренінг позитивно вплинув на показники співробітників.